File: app/src/main/java/com/app/ralaunch/utils/CoreCLRConfig.java
package com.app.ralaunch.utils;

import android.content.Context;
import com.app.ralaunch.data.SettingsManager;

/**
 * CoreCLR 配置工具类
 *
 * 负责将用户设置的 CoreCLR 配置应用到环境变量中，
 * 这些环境变量会在 .NET 运行时启动时被读取。
 *
 * 支持的配置项：
 * - GC 配置：Server GC、Concurrent GC、Heap Count、Retain VM
 * - JIT 配置：Tiered Compilation、Quick JIT、Optimize Type
 */
public class CoreCLRConfig {
    private static final String TAG = "CoreCLRConfig";

    /**
     * 应用 CoreCLR 配置到 native 层
     * 此方法需要在启动 .NET 运行时之前调用
     *
     * @param context Android Context
     */
    public static void applyConfig(Context context) {
        SettingsManager settings = SettingsManager.getInstance(context);
        // GC 配置
        applyGCConfig(settings);
        // JIT 配置
        applyJITConfig(settings);
    }

    /**
     * 应用 GC 配置
     */
    private static void applyGCConfig(SettingsManager settings) {
        boolean serverGC = settings.isServerGC();
        setNativeEnv("DOTNET_gcServer", serverGC ? "1" : "0");

        boolean concurrentGC = settings.isConcurrentGC();
        setNativeEnv("DOTNET_gcConcurrent", concurrentGC ? "1" : "0");

        // GC Heap Count
        String heapCount = settings.getGCHeapCount();
        if (!"auto".equals(heapCount)) {
            setNativeEnv("DOTNET_GCHeapCount", heapCount);
        }

        boolean retainVM = settings.isRetainVM();
        setNativeEnv("DOTNET_GCRetainVM", retainVM ? "1" : "0");
    }

    /**
     * 应用 JIT 配置
     */
    private static void applyJITConfig(SettingsManager settings) {
        boolean tieredCompilation = settings.isTieredCompilation();
        setNativeEnv("DOTNET_TieredCompilation", tieredCompilation ? "1" : "0");

        boolean quickJIT = settings.isQuickJIT();
        if (tieredCompilation) {
            setNativeEnv("DOTNET_TC_QuickJit", quickJIT ? "1" : "0");
        }

        int optimizeType = settings.getJitOptimizeType();
        setNativeEnv("DOTNET_JitOptimizeType", String.valueOf(optimizeType));
    }

    /**
     * 通过 JNI 设置 native 层的环境变量
     * 这个方法会调用 C++ 的 setenv 函数
     *
     * @param key 环境变量名
     * @param value 环境变量值
     */
    private static void setNativeEnv(String key, String value) {
        try {
            nativeSetEnv(key, value);
        } catch (UnsatisfiedLinkError e) {
        }
    }

    /**
     * 获取当前 CoreCLR 配置的摘要信息
     *
     * @param context Android Context
     * @return 配置摘要字符串
     */
    public static String getConfigSummary(Context context) {
        SettingsManager settings = SettingsManager.getInstance(context);
        StringBuilder sb = new StringBuilder();

        sb.append("CoreCLR 配置摘要:\n");
        sb.append("  GC:\n");
        sb.append("    Server GC: ").append(settings.isServerGC() ? "启用" : "关闭").append("\n");
        sb.append("    Concurrent GC: ").append(settings.isConcurrentGC() ? "启用" : "关闭").append("\n");
        sb.append("    Heap Count: ").append(settings.getGCHeapCount()).append("\n");
        sb.append("    Retain VM: ").append(settings.isRetainVM() ? "启用" : "关闭").append("\n");
        sb.append("  JIT:\n");
        sb.append("    Tiered Compilation: ").append(settings.isTieredCompilation() ? "启用" : "关闭").append("\n");
        sb.append("    Quick JIT: ").append(settings.isQuickJIT() ? "启用" : "关闭").append("\n");

        int optimizeType = settings.getJitOptimizeType();
        String optimizeTypeName;
        switch (optimizeType) {
            case 1: optimizeTypeName = "体积优先"; break;
            case 2: optimizeTypeName = "速度优先"; break;
            default: optimizeTypeName = "混合"; break;
        }
        sb.append("    Optimize Type: ").append(optimizeTypeName).append("\n");

        return sb.toString();
    }

    /**
     * Native 方法：设置环境变量
     * 这个方法在 main.cpp 或 netcorehost_launcher.cpp 中实现
     */
    private static native void nativeSetEnv(String key, String value);
}

File: app/src/main/java/com/app/ralaunch/core/platform/runtime/dotnet/CoreCLRConfig.kt
package com.app.ralaunch.core.platform.runtime.dotnet

import android.content.Context
import com.app.ralaunch.core.platform.runtime.EnvVarsManager
import com.app.ralaunch.core.config.AppConfig
import org.koin.java.KoinJavaComponent

/**
 * CoreCLR 配置工具类
 *
 * 负责将用户设置的 CoreCLR 配置应用到环境变量中，
 * 这些环境变量会在 .NET 运行时启动时被读取。
 *
 * 支持的配置项：
 * - GC 配置：Server GC、Concurrent GC、Heap Count、Retain VM
 * - JIT 配置：Tiered Compilation、Quick JIT、Optimize Type
 */
object CoreCLRConfig {
    private const val TAG = "CoreCLRConfig"

    /**
     * 应用 CoreCLR 配置到 native 层
     * 此方法需要在启动 .NET 运行时之前调用
     *
     * @param context Android Context
     */
    fun applyConfigAndInitHooking() {
        val context: Context = KoinJavaComponent.get(Context::class.java)
        EnvVarsManager.quickSetEnvVars(
            // 应用 GC 配置
            "DOTNET_gcServer" to if (AppConfig.c.serverGC) "1" else "0",
            "DOTNET_gcConcurrent" to if (AppConfig.c.concurrentGC) "1" else "0",
            "DOTNET_GCHeapCount" to AppConfig.c.gcHeapCount.takeIf { it != "auto" },
            "DOTNET_GCRetainVM" to if (AppConfig.c.retainVM) "1" else "0",

            // 应用 JIT 配置
            "DOTNET_TieredCompilation" to if (AppConfig.c.tieredCompilation) "1" else "0",
            "DOTNET_TC_QuickJit" to if (AppConfig.c.tieredCompilation && AppConfig.c.quickJIT) "1" else "0",
            "DOTNET_JitOptimizeType" to AppConfig.c.jitOptimizeType.toString(),

            // 应用日志配置
            "COMPlus_DebugWriteToStdErr" to "1",
            "COREHOST_TRACE" to if (AppConfig.c.verboseLogging) "1" else "0",
            "COREHOST_TRACEFILE" to
                    if (AppConfig.c.verboseLogging)
                        context.getExternalFilesDir(null)?.resolve("corehost_trace.txt")?.absolutePath
                    else
                        null,

            // 应用版本前滚策略
            "DOTNET_ROLL_FORWARD" to "LatestMajor",
            "DOTNET_ROLL_FORWARD_ON_NO_CANDIDATE_FX" to "2",
            "DOTNET_ROLL_FORWARD_TO_PRERELEASE" to "1",
        )

        if (AppConfig.c.verboseLogging) {
            CoreHostHooks.initTraceHooks()
        }
    }

    /**
     * 获取当前 CoreCLR 配置的摘要信息
     *
     * @param context Android Context
     * @return 配置摘要字符串
     */
    fun getConfigSummary(context: Context?): String {
        val sb = StringBuilder()

        sb.append("CoreCLR 配置摘要:\n")
        sb.append("  GC:\n")
        sb.append("    Server GC: ").append(if (AppConfig.c.serverGC) "启用" else "关闭")
            .append("\n")
        sb.append("    Concurrent GC: ").append(if (AppConfig.c.concurrentGC) "启用" else "关闭")
            .append("\n")
        sb.append("    Heap Count: ").append(AppConfig.c.gcHeapCount).append("\n")
        sb.append("    Retain VM: ").append(if (AppConfig.c.retainVM) "启用" else "关闭")
            .append("\n")
        sb.append("  JIT:\n")
        sb.append("    Tiered Compilation: ")
            .append(if (AppConfig.c.tieredCompilation) "启用" else "关闭").append("\n")
        sb.append("    Quick JIT: ").append(if (AppConfig.c.quickJIT) "启用" else "关闭")
            .append("\n")

        val optimizeType = AppConfig.c.jitOptimizeType
        val optimizeTypeName: String?
        when (optimizeType) {
            1 -> optimizeTypeName = "体积优先"
            2 -> optimizeTypeName = "速度优先"
            else -> optimizeTypeName = "混合"
        }
        sb.append("    Optimize Type: ").append(optimizeTypeName).append("\n")

        return sb.toString()
    }
}
